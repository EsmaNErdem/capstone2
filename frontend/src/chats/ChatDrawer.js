import React, { useRef, useState, useEffect, useContext } from "react";
import UserContext from '../auth/UserContext';
import { Box, SwipeableDrawer, IconButton } from '@mui/material';
import CloseIcon from '@mui/icons-material/Close';
import Chat from "./Chat";
import Alert from "../utilities/Alert";
import BookClubApi from "../api";

/**
 * ChatDrawer Component
 * 
 * Displays a Material UI pop-up drawer for chat between current user and user on profile view.
 * 
 * - Displays input text area for chat text input
 * - Lists previous messages between users
 * - Makes API call to load previous messages on component mount
 * 
 * - Profile ==> ChatDrawer ==> Chat
 */
const ChatDrawer = ({ isOpen, onClose, receiver }) => {
    console.debug("ChatDrawer");

    const drawerRef = useRef();
    const { currentUser } = useContext(UserContext);
    const [messages, setMessages] = useState([]);
    const [websocket, setWebsocket] = useState(false)
    const [error, setError] = useState(null);

    /**
     * Fetches the previous messages between users from backend sends to child component after WebSocket connection is established
     */ 
    useEffect(function getPreviousMessages() {
        console.debug("ChatDrawer useEffect getPreviousMessages");

        const getMessages = async () => {
            try{
                if (isOpen && websocket) {
                    const roomName = `${receiver},${currentUser.username}`
                    const previousMessages = await BookClubApi.getRoomPreviousMessages(roomName);
                    setMessages([...previousMessages]);                 
                }
            } catch (e) {
                console.error("ChatDrawer-previous messages useEffect API call data loading error:", e);
                setError("An error occurred while fetching previous messages.");
            }
        }
        getMessages();
    }, [isOpen, receiver, websocket]);

    return (
        <SwipeableDrawer
            ref={drawerRef} 
            anchor="right"
            open={isOpen}
            onClose={onClose}
            onOpen={() => {}}
        >
            <Box
                sx={{
                    width: '30rem',
                    padding: '10px', 
                    height: '100vh',
                    display: 'flex', 
                    flexDirection: 'column', 
                    justifyContent: "center"
                }}
                role="presentation"
                onKeyDown={(e) => {
                    if (e.key === 'Escape') {
                        onClose();
                    }
                }}
                data-testid="box"
            >
                <IconButton onClick={onClose} sx={{ position: 'absolute', top: 15, right: '27rem' }}>
                    <CloseIcon />
                </IconButton>
                {error ? <Alert type="danger" messages={[error]} />: null}
                <Chat 
                    isOpen={isOpen} 
                    receiver={receiver}
                    prevMessages={messages}
                    setWebsocket={setWebsocket}
                />

            </Box>
        </SwipeableDrawer>
    );
}

export default ChatDrawer;