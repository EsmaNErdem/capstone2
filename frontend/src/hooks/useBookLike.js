import { useState, useEffect, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user book like
 * This hook provides functionality to manage user likes on books, including retrieving initial like status and count,
 * updating the like status and count, and handling user interactions to like or unlike a book.
 * 
 * -Utilizes useContext data to send user like to database
 * -Creates state to update book like count and user book like status
 * -Use useEffect to avoid to recalculations
 * 
 * Returns user like status, book like count states and handleLike function
 */
const useBookLike = (bookId, initialLikes, bookData={}) => {
    console.debug("useBookLike"); 

    const { hasLikedBook, likeBook } = useContext(UserContext);
    const [liked, setLiked] = useState();
    const [likes, setLikes] = useState();
    const [error, setError] = useState(null);

    // Helper function to convert "undefined" or "null" strings to actual undefined
    const convertUndefined = (value) => (value === "undefined" || value === null ? undefined : value);

    // Convert undefined/null strings in the bookData object
    const convertedBookData = Object.fromEntries(
        Object.keys(bookData).map((key) => [key, convertUndefined(bookData[key])])
    );
        
    /**By using the useEffect, the liked status is only recalculated when the id avoiding unnecessary recalculations on every render.  */
    useEffect(() => {
      setLiked(hasLikedBook(bookId));
      setLikes(initialLikes)
    }, [bookId, initialLikes]);
    
    /**
     * Send user like to database
     * Updates the like count and status based on user actions.
     */ 
    const handleLikeBook = async () => {
        try {
            setError(null);
            if (liked) {
                const unlikedBookId = await likeBook(bookId);
                if (unlikedBookId === bookId) {
                    setLikes((likes) => likes - 1);
                    setLiked(false);
                }
            } else {
                const likedBookId = await likeBook(bookId, convertedBookData);
                if (likedBookId === bookId) {
                    setLikes((likes) => likes + 1);
                    setLiked(true);
                }
            }
        } catch (error) {
            setError('Error liking book.');
            console.error('Error handling book like:', error);
        }
    };
    
  return { liked, likes, error, handleLikeBook };
};

export default useBookLike;