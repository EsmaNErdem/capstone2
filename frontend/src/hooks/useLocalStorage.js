import { useState, useEffect } from "react";

/**
 * Custom Hook to save user info like token and username into localStorage
 * 
 * Creates a user state and looks in localStorage for current user info, if cannot find it, assign null
 * When user info updates, useEffect either cleans it or updates localStorage
 * 
 * Retunrs [user, setUser]
 */

const useLocalStorage = (localStorageKey, initialValue = null) => {
    let initialUserValue
    if(localStorage.getItem(localStorageKey)) {
        try {
            initialUserValue = JSON.parse(localStorage.getItem(localStorageKey))
        } catch (e) {
            console.error("useLocalStorage custom hook: problem loading localStorage data", e);
        }
    } else {
        initialUserValue = initialValue
    }

    const [user, setUser] = useState(initialUserValue)

    useEffect(() => {
        if(user === null) {
            localStorage.removeItem(localStorageKey)
        } else {
            localStorage.setItem(localStorageKey, JSON.stringify(user))
        }
    }, [localStorageKey, user])

    return [user, setUser]
}

export default useLocalStorage;