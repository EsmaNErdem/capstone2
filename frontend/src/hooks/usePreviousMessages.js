import { useState, useEffect } from "react";
import BookClubApi from "../api";

/** Custom Hook for fetching previous messages between receiver and current user
 * This hook makes API call to backend to retrieve related data and set messaegs state
 * 
 * Returns error and messages
 */
const usePreviousMessages = (receiver, currentUserName, websocket, isOpen) => {
    const [error, setError] = useState(null);
    const [messages, setMessages] = useState([]);

    /**
     * Fetches the previous messages between users from backend sends to child component after WebSocket connection is established
     */ 
    useEffect(function getPreviousMessages() {
        console.debug("usePreviousMessages hook useEffect getPreviousMessages");
    
        const getMessages = async () => {
            try{
                if (isOpen && websocket) {
                    const roomName = `${receiver},${currentUserName}`
                    const previousMessages = await BookClubApi.getRoomPreviousMessages(roomName);
                    setMessages([...previousMessages]);                 
                }
            } catch (e) {
                console.error("usePreviousMessages useEffect API call data loading error:", e);
                setError("An error occurred while fetching previous messages.");
            }
        }
        getMessages();
    }, [isOpen, receiver, websocket]);

    return { error, messages }
}

export default usePreviousMessages;