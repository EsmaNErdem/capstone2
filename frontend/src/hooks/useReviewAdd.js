import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user review submission on books/:id route components
 * This hook provides functionality to manage user review submission and updateting bookReviews state
 * 
 * -Utilizes useContext data to send user review to database
 * -Updates book review
 * 
 * Returns error and addBookReview function
 */
const useReviewAdd = (setBookReviews, bookData={}, errorHandlingState) => {
    console.debug("useReviewAdd"); 

    const { addUserReview } = useContext(UserContext);
    const [error, setError] = useState(null);

    // add current user's book review, function is called in BookReviewDrawer
    const addBookReview = async ({ review }) => {
        try {
            console.log({book: bookData, review})
            const newReview = await addUserReview({book: bookData, review});
            setBookReviews(r => [{...newReview, id:bookData.id}, ...r])
            return newReview
        } catch (error) {
            errorHandlingState(false)
            setError("Error adding book review.")
            console.error("Error adding book review:", error);
        }
    } 
    
    return { error, addBookReview }

}

export default useReviewAdd