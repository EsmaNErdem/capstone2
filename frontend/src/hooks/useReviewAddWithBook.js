import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user review submission for /reviews route components
 * This hook provides functionality to manage user review submission and updateting bookReviews state
 * 
 * -Utilizes useContext data to send user review to database
 * -Updates book review
 * 
 * Returns error and addBookReview function
 */
const useReviewAddWithBook = (setBookReviews, errorHandlingState) => {
    console.debug("useReviewAddWithBook"); 

    const { addUserReview } = useContext(UserContext);
    const [error, setError] = useState(null);

    // add current user's book review, function is called in BookReviewDrawer
    const addBookReview = async (review, bookData={}) => {
        try {
            const convertedBookData = convertNullInObject(bookData);
            const newReview = await addUserReview({book: convertedBookData, review});
            setBookReviews(r => [{...newReview, book_id:bookData.id, title:bookData.title, author:bookData.author, cover:bookData.cover, category:bookData.category, reviewLikeCount:0}, ...r])
            return newReview
        } catch (error) {
            errorHandlingState(false)
            setError("Error adding book review.")
            console.error("Error adding book review:", error);
        }
    } 
    
    return { error, addBookReview }

}

export default useReviewAddWithBook

const convertNullInObject = (obj) => {
    const convertedObj = {};
    for (const key in obj) {
      convertedObj[key] = obj[key] === null ? "" : obj[key];
    }
    return convertedObj;
  };