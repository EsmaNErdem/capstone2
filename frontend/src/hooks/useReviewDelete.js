import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user review delete
 * This hook provides functionality to manage user review delete and updateting bookReviews state
 * 
 * -Utilizes useContext data to delete user review from database
 * -Updates book review
 * 
 * Returns error and addBookReview function
 */
const useReviewDelete = (setBookReviews) => {
    console.debug("useReviewDelete"); 

    const { deleteUserReview } = useContext(UserContext);
    const [error, setError] = useState(null);

    // deletes current user's book review, function is called in BookReview
    const deleteBookReview = async (reviewId, bookId) => {
        try {
            const deletedReviewId = await deleteUserReview(reviewId);
            if(reviewId === +deletedReviewId){
                // Update the state to remove the deleted review
                setBookReviews((prevBookReviews) => {
                    const updatedBookReviews = prevBookReviews.filter(review => review.reviewId !== +deletedReviewId);
                    return updatedBookReviews;
                });
            }
        } catch (error) {
            setError("Error deleting book review.")
            console.error("Error deleting book review:", error);
        }
    } 
    
    return { error, deleteBookReview }

}

export default useReviewDelete