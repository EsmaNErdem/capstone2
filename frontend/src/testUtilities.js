import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
