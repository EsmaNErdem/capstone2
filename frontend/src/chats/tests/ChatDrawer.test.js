import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, act } from '@testing-library/react';
import ChatDrawer from '../ChatDrawer';
import BookClubApi from '../../api';
import * as usePreviousMessages from '../../hooks/usePreviousMessages';

const mockMessages = [
    {
        name: 'User1',
        text: 'Rest1', 
        date: new Date().toISOString(),
    },
    {
        name: 'User2',
        text: 'test2', 
        date: new Date().toISOString(),
    }
]

jest.mock("../../api", () => ({
    __esModule: true,
    default: {
        getRoomPreviousMessages: jest.fn(),
    },
}));
  

beforeEach(() => {
  jest.spyOn(usePreviousMessages, 'default').mockImplementation(() => ({
    messages: mockMessages,
    error: null,
  }));
});

afterEach(() => {
    jest.clearAllMocks();
});

test('renders without crashing', async () => {
  BookClubApi.getRoomPreviousMessages.mockImplementationOnce(() => (mockMessages));

  await act(async () => {
    render(
      <MemoryRouter>
        <UserProvider>
          <ChatDrawer />
        </UserProvider>
      </MemoryRouter>
    );
  });

});

test('it renders and matches with snapshot', async () => {
  BookClubApi.getRoomPreviousMessages.mockImplementationOnce(() => (mockMessages));

  await act(async () => {
    ({ asFragment } = render(
      <MemoryRouter>
        <UserProvider>
          <ChatDrawer />
        </UserProvider>
      </MemoryRouter>
    ));
  });

  expect(asFragment()).toMatchSnapshot();
});