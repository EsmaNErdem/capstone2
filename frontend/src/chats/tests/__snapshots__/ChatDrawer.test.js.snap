// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="PrivateSwipeArea-root css-x15wq9"
    style="width: 20px;"
  />
</DocumentFragment>
`;
